torch/csrc/jit/import_source.cpp

```diff
@@ -0,0 +1,194 @@
+      return std::make_shared<MethodValue>(shared_from_this(), *m);
+    } else {
+      throw ErrorReport(loc) << "unknown attr: " << field;
+    }
+  }
+
+ private:
+  std::shared_ptr<Module> module;
+};
+
+struct OpsValue : public SugaredValue {
+  OpsValue(size_t version) : version_(version) {}
+  std::string kind() const override {
+    return "ops";
+  }
+  std::shared_ptr<SugaredValue> attr(
+      const SourceRange& loc,
+      Method& m,
+      const std::string& field) override {
+    return std::make_shared<BuiltinModule>(field, version_);
+  }
+  size_t version_;
+};
+
+struct ConstantValue : public SugaredValue {
+  ConstantValue(IValue value) : value_(std::move(value)) {}
+  IValue value_;
+  std::string kind() const override {
+    return "constant";
+  }
+  Value* asValue(const SourceRange& loc, Method& m) override {
+    return m.graph()->insertConstant(value_);
+  }
+};
+
+// This value maps attributes CONSTANTS.c0 CONSTANTS.c1 to entries
+// in the 'constants' vector. This table is will be stored in a container format
+// and given to the import_method when restoring the code.
+struct ConstantTableValue : public SugaredValue {
+  ConstantTableValue(ArrayRef<at::Tensor> constants) : constants_(constants) {}
+  std::string kind() const override {
+    return "CONSTANTS";
+  }
+  // select an attribute on it, e.g. `this.field`
+  std::shared_ptr<SugaredValue> attr(
+      const SourceRange& loc,
+      Method& m,
+      const std::string& field) override {
+    const char* field_s = field.c_str();
+    char* end;
+    int64_t offset = std::strtoll(field_s + 1, &end, 10);
+    if (field.size() < 2 || *end != 0)
+      throw ErrorReport(loc) << "invalid constant specifier: " << field;
+    if (offset < 0 || size_t(offset) >= constants_.size()) {
+      throw ErrorReport(loc) << "constant index " << offset
+                             << " is out of bounds (constant table has "
+                             << constants_.size() << " entries).";
+    }
+    Value* value = m.graph()->insertConstant(constants_[offset], nullptr, loc);
+    return std::make_shared<SimpleValue>(value);
+  }
+
+ private:
+  ArrayRef<at::Tensor> constants_;
+};
+
+// Helper that contains the state for a parsing a TorchScript source string.
+struct SourceImporter {
+  SourceImporter(
+      const std::string& src,
+      const std::vector<at::Tensor>& constant_table)
+      : parser_(src), constant_table_(constant_table) {
+    const auto version = parseVersionNumber();
+    env_ = {
+        {"torch", std::make_shared<BuiltinModule>("aten", version)},
+        {"ops", std::make_shared<OpsValue>(version)},
+        {"CONSTANTS", std::make_shared<ConstantTableValue>(constant_table)},
+        {"fork", std::make_shared<ForkValue>()},
+        {"annotate", std::make_shared<AnnotateValue>()},
+        {"inf",
+         std::make_shared<ConstantValue>(
+             std::numeric_limits<double>::infinity())},
+        {"nan",
```

Comment:
Why was "nan" added here?

Reply:
no clue. It may have been a bad merge :/